File: addon.py
from stremio_addon import addon, catalogs, metas, streams
from stremio_addon.server import run
import requests
from bs4 import BeautifulSoup
import base64
import re

LATANIME_URL = "https://latanime.org/"

MANIFEST = {
    "id": "community.python.latanime",
    "version": "1.0.0",
    "name": "LatAnime (Python)",
    "description": "Addon en Python para ver animes desde LatAnime.org",
    "resources": [
        "catalog",
        "meta",
        "stream"
    ],
    "types": ["series"],
    "catalogs": [
        {
            "type": "series",
            "id": "latanime-animes",
            "name": "Animes"
        }
    ],
    "idPrefixes": ["latanime:"]
}

class LatAnimeAddon(addon.Addon):
    def __init__(self):
        super(LatAnimeAddon, self).__init__(MANIFEST)

    @addon.handler(catalogs.Find)
    def find_catalogs(self, query, **kwargs):
        if query.type == 'series' and query.id == 'latanime-animes':
            try:
                response = requests.get(LATANIME_URL)
                response.raise_for_status()
                soup = BeautifulSoup(response.text, 'html.parser')
                
                meta_previews = []
                for article in soup.select('div.main-col article'):
                    title_tag = article.select_one('h3.title a')
                    poster_tag = article.select_one('div.poster img')
                    
                    if title_tag and poster_tag:
                        title = title_tag.text
                        link = title_tag['href']
                        poster = poster_tag['src']
                        
                        try:
                            slug = link.split('/')[-2]
                            meta_previews.append(metas.MetaPreview(
                                id="latanime:" + slug,
                                type="series",
                                name=title,
                                poster=poster
                            ))
                        except IndexError:
                            continue
                
                return catalogs.Catalog(meta_previews=meta_previews)
            except requests.exceptions.RequestException as e:
                print(f"Error fetching catalog: {e}")
                return None
        return None

    @addon.handler(metas.Find)
    def find_metas(self, query, **kwargs):
        if query.type == 'series' and query.id.startswith('latanime:'):
            slug = query.id.split(':')[1]
            anime_url = f"{LATANIME_URL}animes/{slug}"
            
            try:
                response = requests.get(anime_url)
                response.raise_for_status()
                soup = BeautifulSoup(response.text, 'html.parser')

                name = soup.select_one('h1.title').text
                poster = soup.select_one('div.poster img')['src']
                description = soup.select_one('div.sinopsis').text.strip()

                videos = []
                for li in soup.select('div.episodes li'):
                    link_tag = li.select_one('a')
                    if link_tag:
                        episode_link = link_tag['href']
                        episode_title = link_tag.text
                        
                        try:
                            episode_slug = episode_link.split('/')[-2]
                            episode_number_match = re.search(r'\d+$', episode_slug)
                            if episode_number_match:
                                episode_number = int(episode_number_match.group())
                                videos.append(metas.Video(
                                    id=f"latanime:{episode_slug}",
                                    title=episode_title,
                                    season=1,
                                    episode=episode_number,
                                    released="2024-01-01T00:00:00.000Z" # Placeholder
                                ))
                        except (IndexError, ValueError):
                            continue
                
                videos.reverse()

                meta_obj = metas.Meta(
                    id=query.id,
                    type='series',
                    name=name,
                    poster=poster,
                    description=description,
                    videos=videos
                )
                return [meta_obj]

            except requests.exceptions.RequestException as e:
                print(f"Error fetching meta: {e}")
                return []
        return []

    @addon.handler(streams.Find)
    def find_streams(self, query, **kwargs):
        if query.type == 'series' and query.id.startswith('latanime:'):
            episode_slug = query.id.split(':')[1]
            episode_url = f"{LATANIME_URL}ver/{episode_slug}"

            try:
                response = requests.get(episode_url)
                response.raise_for_status()
                soup = BeautifulSoup(response.text, 'html.parser')

                stream_results = []
                for a_tag in soup.select('a.play-video[data-player]'):
                    encoded_url = a_tag.get('data-player')
                    if encoded_url:
                        try:
                            decoded_url = base64.b64decode(encoded_url).decode('utf-8')
                            server_name = a_tag.text.strip()
                            stream_results.append(streams.Stream(
                                title=server_name,
                                url=decoded_url
                            ))
                        except (base64.binascii.Error, UnicodeDecodeError) as e:
                            print(f"Failed to decode base64 URL: {encoded_url}, {e}")
                
                return stream_results

            except requests.exceptions.RequestException as e:
                print(f"Error fetching streams: {e}")
                return []
        return []

if __name__ == '__main__':
    run(LatAnimeAddon())

File: addon.js
const { addonBuilder } = require("stremio-addon-sdk");
const axios = require("axios");
const cheerio = require("cheerio");

const LATANIME_URL = "https://latanime.org/";

// Manifest
const manifest = {
    "id": "community.stremio.latanime",
    "version": "1.0.0",
    "name": "LatAnime",
    "description": "Addon para ver animes desde LatAnime.org",
    "resources": [
        "catalog",
        "meta",
        "stream"
    ],
    "types": ["series"],
    "catalogs": [
        {
            "type": "series",
            "id": "latanime-animes",
            "name": "Animes"
        }
    ],
    "idPrefixes": [ "latanime:" ]
};

const builder = new addonBuilder(manifest);

// Catalog Handler
builder.defineCatalogHandler(async ({type, id, extra}) => {
    console.log("Request for catalog: " + type + " " + id);
    let metas = [];

    if (type === 'series' && id === 'latanime-animes') {
        try {
            const response = await axios.get(LATANIME_URL);
            const $ = cheerio.load(response.data);

            $('div.main-col article').each((i, element) => {
                const title = $(element).find('h3.title a').text();
                const poster = $(element).find('div.poster img').attr('src');
                const link = $(element).find('h3.title a').attr('href');

                // Extraer el slug de la URL para usarlo como ID
                const urlParts = link.split('/');
                const slug = urlParts[urlParts.length - 2];

                if (title && poster && slug) {
                    metas.push({
                        id: "latanime:" + slug,
                        type: "series",
                        name: title,
                        poster: poster
                    });
                }
            });
        } catch (error) {
            console.error("Error fetching catalog:", error);
            // Devolver un catálogo vacío en caso de error
            return Promise.resolve({ metas: [] });
        }
    }

    return Promise.resolve({ metas: metas });
});

// Meta Handler
builder.defineMetaHandler(async ({type, id}) => {
    console.log("Request for meta: " + type + " " + id);

    const [prefix, slug] = id.split(':');
    const animeUrl = `${LATANIME_URL}animes/${slug}`;

    try {
        const response = await axios.get(animeUrl);
        const $ = cheerio.load(response.data);

        const name = $('h1.title').text();
        const poster = $('div.poster img').attr('src');
        const synopsis = $('div.sinopsis').text().trim();
        
        const videos = [];
        $('div.episodes li').each((i, element) => {
            const episodeLink = $(element).find('a').attr('href');
            const episodeTitle = $(element).find('a').text();
            
            const episodeUrlParts = episodeLink.split('/');
            const episodeSlug = episodeUrlParts[episodeUrlParts.length - 2];
            const episodeNumber = parseInt(episodeSlug.match(/\d+$/)[0], 10);

            if (episodeSlug && !isNaN(episodeNumber)) {
                videos.push({
                    id: `latanime:${episodeSlug}`,
                    title: episodeTitle,
                    season: 1, // LatAnime doesn't seem to have season numbers, so we'll default to 1
                    episode: episodeNumber,
                    released: new Date() // Placeholder for release date
                });
            }
        });

        // Reverse the videos array so that episode 1 is first
        const meta = {
            id: id,
            type: 'series',
            name: name,
            poster: poster,
            synopsis: synopsis,
            videos: videos.reverse()
        };

        return Promise.resolve({ meta: meta });

    } catch (error) {
        console.error("Error fetching meta:", error);
        return Promise.reject(new Error("Failed to get meta for " + id));
    }
});


// Stream Handler
builder.defineStreamHandler(async ({type, id}) => {
    console.log("Request for streams: " + type + " " + id);

    const [prefix, episodeSlug] = id.split(':');
    const episodeUrl = `${LATANIME_URL}ver/${episodeSlug}`;

    try {
        const { data } = await axios.get(episodeUrl);
        const $ = cheerio.load(data);

        let streams = [];

        // --- METHOD 1: Find the `var video = [...]` script block ---
        const scriptContent = $('script:contains("var video =")').html();
        if (scriptContent) {
            const videoArrayMatch = scriptContent.match(/var video = (\s*[^;]+);/);
            if (videoArrayMatch && videoArrayMatch[1]) {
                const videoArrayStr = videoArrayMatch[1];
                const urlRegex = /file:\s*"(.*?)"/g;
                let match;
                while ((match = urlRegex.exec(videoArrayStr)) !== null) {
                    streams.push({
                        title: "Server (JS)",
                        url: match[1]
                    });
                }
            }
        }

        // --- METHOD 2: Use the Player API if Method 1 fails ---
        if (streams.length === 0) {
            const dataId = $('div#Player').attr('data-id');
            if (dataId) {
                console.log("Falling back to API method with data-id:", dataId);
                // We need to send the data as application/x-www-form-urlencoded
                const params = new URLSearchParams();
                params.append('id', dataId);

                const apiResponse = await axios.post(`${LATANIME_URL}api/player/`, params);
                
                if (apiResponse.data && apiResponse.data.player) {
                    const playerHtml = apiResponse.data.player;
                    const $ = cheerio.load(playerHtml);
                    $('iframe').each((i, el) => {
                        const streamUrl = $(el).attr('src');
                        if (streamUrl) {
                            streams.push({
                                title: "Server (API)",
                                url: streamUrl
                            });
                        }
                    });
                }
            }
        }

        if (streams.length === 0) {
            throw new Error("No stream URLs found on page: " + episodeUrl);
        }

        return Promise.resolve({ streams: streams });

    } catch (error) {
        console.error("Error fetching streams:", error);
        return Promise.reject(new Error("Failed to get streams for " + id));
    }
});

module.exports = builder.getInterface();